linux/library/src/text_input_plugin.cc

```diff
@@ -0,0 +1,161 @@
+  active_model_->AddCharacter(static_cast<char>(code_point));
+  CallPlatformCallback(active_model_->GetState());
+}
+
+void TextInputPlugin::KeyboardHook(GLFWwindow *window, int key, int scancode,
+                                   int action, int mods) {
+  if (active_model_ == nullptr) {
+    return;
+  }
+  if (action == GLFW_PRESS || action == GLFW_REPEAT) {
+    switch (key) {
+      case GLFW_KEY_LEFT:
+        if (active_model_->MoveCursorBack()) {
+          CallPlatformCallback(active_model_->GetState());
+        }
+        break;
+      case GLFW_KEY_RIGHT:
+        if (active_model_->MoveCursorForward()) {
+          CallPlatformCallback(active_model_->GetState());
+        }
+        break;
+      case GLFW_KEY_END:
+        active_model_->MoveCursorToEnd();
+        CallPlatformCallback(active_model_->GetState());
+        break;
+      case GLFW_KEY_HOME:
+        active_model_->MoveCursorToBeginning();
+        CallPlatformCallback(active_model_->GetState());
+        break;
+      case GLFW_KEY_BACKSPACE:
+        if (active_model_->Backspace()) {
+          CallPlatformCallback(active_model_->GetState());
+        }
+        break;
+      case GLFW_KEY_DELETE:
+        if (active_model_->Delete()) {
+          CallPlatformCallback(active_model_->GetState());
+        }
+        break;
+      default:
+        break;
+    }
+  }
+}
+
+TextInputPlugin::TextInputPlugin(PlatformCallback platform_callback)
+    : Plugin(kChannelName, platform_callback, false), active_model_(nullptr) {}
+
+TextInputPlugin::~TextInputPlugin() {}
+
+Json::Value TextInputPlugin::HandlePlatformMessage(const Json::Value &message) {
+  Json::Value method = message[kMethodKey];
+  if (method.isNull()) {
+    std::cerr << "No text input method declaration" << std::endl;
+    return Json::nullValue;
+  }
+  std::string method_str = method.asString();
+
+  // These methods are no-ops.
+  if (method_str.compare(kShowMethod) == 0 ||
+      method_str.compare(kHideMethod) == 0) {
+    return Json::nullValue;
+  }
+
+  // Clear client method takes no args. Every following method requires args.
+  if (method_str.compare(kClearClientMethod) == 0) {
+    active_model_ = nullptr;
+    return Json::nullValue;
+  }
+
+  Json::Value args = message[kArgumentsKey];
+  if (args.isNull()) {
+    std::cerr << "Method invoked without args" << std::endl;
+    return Json::nullValue;
+  }
+
+  if (method_str.compare(kSetClientMethod) == 0) {
+    // TODO(awdavies): There's quite a wealth of arguments supplied with this
+    // method, and they should be inspected/used.
+    Json::Value client_id_json = args[0];
+    if (client_id_json.isNull()) {
+      std::cerr << "Could not set client, ID is null." << std::endl;
+      return Json::nullValue;
+    }
+    int client_id = client_id_json.asInt();
+    if (input_models_.find(client_id) == input_models_.end()) {
+      input_models_.insert(std::make_pair(
+          client_id, std::make_unique<TextInputModel>(client_id)));
```

Comment:
I think we should, yes.  However inflating text ID fields should be fixed once we send back valid callback responses via HandlePlatformMessage (instead of returning null).